README.md
# Mojolicious-Plugin-SetUserGroup
Mojolicious::Plugin::SetUserGroup - A Mojolicious plugin to set unprivileged credentials


Changes
{{$NEXT}}

1.000     2018-02-19 14:51:49 EST
  - Use Unix::Groups::FFI to set supplemental groups with initgroups()
  - Only skip changing credentials if real user and group are already correct, not effective user and group
    (this usually does not matter, but an effective-only credential change can be reverted)

0.005     2016-08-01 02:42:04 EDT
  - Further fixes to error handling for credential change failure
  - Fix to terminate prefork server properly on credential change failure in recent Mojolicious
  - Don't attempt to set user and group if they are both already correct
  - Add warning about morbo compatibility to docs

0.004     2016-04-04 21:36:54 EDT
  - Fix and improve syscall error handling

0.003     2015-11-06 23:28:34 EST
  - Switch installer to Module::Build::Tiny
  - Fix test failure when no group exists with the same name as the current user

0.002     2015-04-27 10:29:29 EDT
  - Die on plugin register with nonexistent user or group (leejo)

0.001     2015-04-26 20:23:40 EDT
  - First release


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.036.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Mojolicious plugin to set unprivileged credentials",
  "AUTHOR" => "Dan Book <dbook\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Mojolicious-Plugin-SetUserGroup",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "Mojolicious::Plugin::SetUserGroup",
  "PREREQ_PM" => {
    "List::Util" => "1.33",
    "Mojolicious" => "6.0",
    "POSIX" => 0,
    "Unix::Groups" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => "0.88"
  },
  "VERSION" => "0.003",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "List::Util" => "1.33",
  "Mojolicious" => "6.0",
  "POSIX" => 0,
  "Test::More" => "0.88",
  "Unix::Groups" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Mojolicious/Plugin/SetUserGroup.pm
package Mojolicious::Plugin::SetUserGroup;
use Mojo::Base 'Mojolicious::Plugin';

use Mojo::IOLoop;
use POSIX qw(getuid getgid setuid setgid);
use Unix::Groups::FFI 'initgroups';
use Carp 'croak';

our $VERSION = '1.001';

sub register {
	my ($self, $app, $conf) = @_;
	my $user = $conf->{user};
	my $group = $conf->{group} // $user;
	
	return $self unless defined $user;
	
	# Make sure desired user and group exist
	$! = 0;
	unless (defined(getpwnam $user)) {
		croak _error($app, qq{Failed to retrieve user "$user": $!}) if $!;
		croak _error($app, qq{User "$user" does not exist});
	}
	$! = 0;
	unless (defined(getgrnam $group)) {
		croak _error($app, qq{Failed to retrieve group "$group": $!}) if $!;
		croak _error($app, qq{Group "$group" does not exist});
	}
	
	Mojo::IOLoop->next_tick(sub { _setusergroup($app, $user, $group) });
}

sub _error {
	my ($app, $error) = @_;
	chomp $error;
	$app->log->fatal($error);
	Mojo::IOLoop->stop if Mojo::IOLoop->is_running;
	return $error;
}

sub _setusergroup {
	my ($app, $user, $group) = @_;
	
	# User and group IDs
	my ($uid, $gid);
	$! = 0;
	unless (defined($uid = getpwnam $user)) {
		die _error($app, qq{Failed to retrieve user "$user": $!}) if $!;
		die _error($app, qq{User "$user" does not exist});
	}
	$! = 0;
	unless (defined($gid = getgrnam $group)) {
		die _error($app, qq{Failed to retrieve group "$group": $!}) if $!;
		die _error($app, qq{Group "$group" does not exist});
	}
	
	# Check if user and group are already correct
	return if getuid() == $uid and getgid() == $gid;
	
	my $rc = setgid($gid);
	unless (defined $rc and $rc == 0) { die _error($app, qq{Can't switch to group "$group": $!}); }
	my $error;
	{ local $@; unless (eval { initgroups($user, $gid); 1 }) { $error = "$!" } }
	if (defined $error) { die _error($app, qq{Can't set supplemental groups for user "$user": $error}); }
	$rc = setuid($uid);
	unless (defined $rc and $rc == 0) { die _error($app, qq{Can't switch to user "$user": $!}); }
}

1;

=head1 NAME

Mojolicious::Plugin::SetUserGroup - Mojolicious plugin to set unprivileged
credentials

=head1 SYNOPSIS

  # Mojolicious
  $self->plugin(SetUserGroup => {user => $user, group => $group});
  
  # Mojolicious::Lite
  plugin SetUserGroup => {user => $user, group => $group};
  
  # Production mode only
  plugin SetUserGroup => {user => $user, group => $group}
    if $self->mode eq 'production';
  
  # Root only
  plugin SetUserGroup => {user => $user, group => $group}
    if $> == 0;

=head1 DESCRIPTION

This plugin is intended to replace the C<setuidgid> functionality of
L<Mojo::Server>. It should be loaded in application startup and it will change
the user and group credentials of the process when L<Mojo::IOLoop> is started,
which occurs in each worker process of a L<Mojo::Server::Prefork> daemon like
L<hypnotoad>.

This allows an application to be started as root so it can bind to privileged
ports such as port 80 or 443, but run worker processes as unprivileged users.
However, if the application is not started as root, it will most likely fail to
change credentials. So, you should only set the user/group when the application
is started as root or a user with the C<CAP_SETUID> and C<CAP_SETGID>
L<capabilities(7)>.

This plugin only works on systems with a concept of Unix users and groups, such
as Linux, OS X, or BSD.

The L<morbo> development server is currently incompatible with this plugin as
the lowered credentials causes the application worker to shut down. Make sure
credential changes do not occur when running your application under morbo,
either by not registering the plugin under that condition, or starting morbo
under the target user and group so no change occurs.

=head1 METHODS

L<Mojolicious::Plugin::SetUserGroup> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones.

=head2 register

  $plugin->register(Mojolicious->new, {user => $user, group => $group});

Install callback to change process credentials on the next L<Mojo::IOLoop>
tick. If option C<user> is undefined or the current user and group are already
correct, no credential change will occur. If option C<group> is undefined but
C<user> is defined, the group will be set to a group matching the user name. If
credential changes fail, an error will be logged and the process will be
stopped.

=head1 AUTHOR

Dan Book, C<dbook@cpan.org>

=head1 CONTRIBUTORS

=over

=item Jan Henning Thorsen (jhthorsen)

=item Lee Johnson (leejo)

=back

=head1 COPYRIGHT AND LICENSE

Copyright 2015, Dan Book.

This library is free software; you may redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<Mojolicious>, L<POSIX>, L<Unix::Groups::FFI>


Build.PL
# This Build.PL for Mojolicious-Plugin-SetUserGroup was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.025
use strict;
use warnings;

my %configure_requires = (
    'Module::Build::Tiny' => '0.034',
);

my %errors = map {
    eval "require $_; $_->VERSION($configure_requires{$_}); 1";
    $_ => $@,
} keys %configure_requires;

if (!grep { $_ } values %errors)
{
    # This section for Mojolicious-Plugin-SetUserGroup was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.015.
    use strict;
    use warnings;

    use 5.010001;
    # use Module::Build::Tiny 0.034;
    Module::Build::Tiny::Build_PL();
}
else
{
    if (not $ENV{PERL_MB_FALLBACK_SILENCE_WARNING})
    {
        warn <<'EOW'
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.


Errors from configure prereqs:
EOW
        . do {
            require Data::Dumper; Data::Dumper->new([ \%errors ])->Indent(2)->Terse(1)->Sortkeys(1)->Dump;
        };

        sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
    }


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.011.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "configure_requires" => {
        "Module::Build::Tiny" => "0.034"
      },
      "dist_abstract" => "Mojolicious plugin to set unprivileged credentials",
      "dist_author" => [
        "Dan Book <dbook\@cpan.org>"
      ],
      "dist_name" => "Mojolicious-Plugin-SetUserGroup",
      "dist_version" => "0.006",
      "license" => "artistic_2",
      "module_name" => "Mojolicious::Plugin::SetUserGroup",
      "recursive_test_files" => 1,
      "requires" => {
        "Mojolicious" => "6.0",
        "POSIX" => 0,
        "Unix::Groups::FFI" => 0,
        "perl" => "5.010001"
      },
      "test_requires" => {
        "File::Spec" => 0,
        "Module::Metadata" => 0,
        "Test::More" => "0.88"
      }
    );


    my %fallback_build_requires = (
      "File::Spec" => 0,
      "Module::Metadata" => 0,
      "Test::More" => "0.88"
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);


    $build->create_build_script;
}
